File: src/ImageProcessor/ImageProcessor/Filters/ThresholdFilter.cs
using System;
using System.Drawing;
using System.Linq;
using System.Text.RegularExpressions;
using ImageProcessor.Models;

namespace ImageProcessor.Filters
{
	public class ThresholdFilter : FilterBase<ThresholdFilterModel>
	{
		public override CommandsLineArg Argument
		{
			get
			{
				return CommandsLineArg.ThresholdFilter;
			}
		}
		public override RawImage Process(ThresholdFilterModel arg, RawImage image)
		{
			var roi = (arg.Roi ?? arg.DefaultRoi).Region;
			var threshold = Convert.ToByte(arg.Threshold * 255);

			for (var y = roi.Y; y < roi.Bottom; y++)
				for (var x = roi.X; x < roi.Right; x++)
					image.SetPixel(x, y, image.ThresholdFilter(x, y, threshold));

			return image;
		}
		public override ThresholdFilterModel Parse(CommandLineArgModel arg)
		{
			return new ThresholdFilterModel
			{
				Roi = RoiModel.Parse(arg),
				Threshold = arg.Parameters
					.Where(param => Regex.IsMatch(param, "^" + ThresholdFilterModel.ThresholdPattern + "$", RegexOptions.IgnoreCase))
					.Select(param => (double?)Convert.ToDouble(param.Substring(param.IndexOf(':') + 1).Trim()))
					.FirstOrDefault() ?? 0.5d
			};
		}
	}
	public class ThresholdFilterModel : IRoiModel
	{
		public RoiModel DefaultRoi
		{
			get;
			set;
		}
		public RoiModel Roi
		{
			get;
			set;
		}
		public const string ThresholdPattern = @"\s*threshold\s*\:\s*(0*|0*1(\.?|\.0*)|0*\.\d+)\s*";
		public double Threshold
		{
			get;
			set;
		}
	}
}

File: src/ImageProcessor/ImageProcessorCpp/thresholdFilter.cpp
#include "thresholdFilter.h"

RawImage* ThresholdFilter::ProcessInput(CommandLineArgModel* arg, RawImage* image)
{
	ThresholdFilterModel* model = (ThresholdFilterModel*)arg->ParsedModel;
	Rectangle roi = model->Roi;
	double threshold = model->Threshold;

	for (int y = roi.Y; y < roi.Bottom; y++)
		for (int x = roi.X; x < roi.Right; x++)
			image->SetPixel(x, y, image->ThresholdFilter(x, y, roundToByte(threshold)));

	return image;
}
bool ThresholdFilter::IsThisArgumentForThisFilter(string arg)
{
	return isMatch("ThresholdFilter", arg);
}
void ThresholdFilter::ParseArgument(CommandLineArgModel* arg)
{
	ThresholdFilterModel* model = new ThresholdFilterModel();
	model->hasRoi = false;
	model->Threshold = 128;

	bool hasThreshold = false;

	for (unsigned int index = 0; index < arg->Parameters->size(); index++)
	{
		if (!model->hasRoi)
		{
			Rectangle roi = parseRoi(arg->Parameters->at(index));
			if (roi.Bottom != -1)
			{
				model->hasRoi = true;
				model->Roi = roi;
			}
		}

		if (!hasThreshold)
		{
			string value = getParamValue("threshold", arg->Parameters->at(index));

			if (value.length())
			{
				hasThreshold = true;
				model->Threshold = roundToByte(stod(trim(value)) * 255);
			}
		}
	}
	
	arg->ParsedModel = model;
}
